src/algebra/order/rearrangement.lean

```diff
@@ -0,0 +1,376 @@
+/-
+Copyright (c) 2021 Mantas Bakšys. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mantas Bakšys
+-/
+import algebra.order.smul
+import group_theory.perm.sign
+import order.monovary
+import tactic.abel
+
+/-!
+# Rearrangement inequality
+
+This file proves the rearrangement inequality.
+
+The rearrangement inequality tells you that for two functions `f g : ι → α`, the sum
+`∑ i, f i * g (σ i)` is maximized over all `σ : perm ι` when `g ∘ σ` monovaries with `f` and
+minimized when `g ∘ σ` antivaries with `f`.
+
+The case for `monotone` pairs of functions over a `linear_order` is deduced.
+
+Note : I'm currently very unsure if this theorem should have it's own file, please feel free to
+comment on this
+-/
+
+variables {ι ι' α β γ : Type*}
+
+section vary
+variables [preorder α] [preorder β] {f : ι → α} {g : ι → β}
+
+open equiv
+
+end vary
+
+namespace finset
+
+/-- Induction principle for `finset` based on the ordering induced by a function. -/
+lemma induction_on_max_value [linear_order α] [decidable_eq ι] (f : ι → α)
+  {p : finset ι → Prop} (s : finset ι) (h0 : p ∅)
+  (step : ∀ a s, a ∉ s → (∀ x ∈ s, f x ≤ f a) → p s → p (insert a s)) : p s :=
+begin
+  induction s using finset.strong_induction_on with s ihs,
+  rcases (s.image f).eq_empty_or_nonempty with hne|hne,
+  { simp only [image_eq_empty] at hne,
+    simp only [hne, h0] },
+  { have H : (s.image f).max' hne ∈ (s.image f), from max'_mem (s.image f) hne,
+    simp only [mem_image, exists_prop] at H,
+    rcases H with ⟨a, has, hfa⟩,
+    rw ← insert_erase has,
+    refine step _ _ (not_mem_erase a s) (λ x hx, _) (ihs _ $ erase_ssubset has),
+    rw hfa,
+    exact le_max' _ _ (mem_image_of_mem _ $  mem_of_mem_erase hx) }
+end
+
+/-- Induction principle for `finset` based on an ordering induced by a pair of functions -/
+lemma finset.induction_on_max' [decidable_eq ι] [linear_order α] [linear_order β] (f : ι → α)
```

Comment:
Duals of this and the above would be nice, and I imagine you plan to move this already!

Reply:
Yup, working on it!